Add API reference and example for USBCDC_ECM class

juhaylinen opened the pull request.

juhaylinen committed:
Add API reference and example for USBCDC_ECM class

Commit:
Edit USBCDC_ECM.md

Edit file to fix note rendering and move example content under example heading.

AnotherButler commented on docs/api/usb/USBCDC_ECM.md:
```diff
@@ -0,0 +1,52 @@
+
+<span class="images">![](https://os.mbed.com/docs/mbed-os/development/mbed-os-api-doxy/class_u_s_b_c_d_c___e_c_m.png)<span>USBCDC_ECM class hierarchy</span></span>
+
+The USBCDC_ECM class emulates an Ethernet interface over USB. You can use the interface to send and receive Ethernet frames over USB.
```
Query: What are the typical use cases for this class? What are the common scenarios in which you'd use it?

juhaylinen replied:
Typical applications are USB to Ethernet adapter and USB Web server that allows users to configure a USB device and to view data using a web browser.  The user needs to add required SW components on top of the USBCDC_ECM class like IP stack and Web server.

AnotherButler commented on docs/api/usb/USBCDC_ECM.md:
````diff
@@ -0,0 +1,52 @@
+4. Click **Capture > Start**.
+5. Click captured packet from source address 12:34:56:78:9a:bc to see the "Hello world" payload.
+
+```C++ TODO
````
Are there plans to make this an importable example? If there's no set plan for this, please add this example to this repository: https://github.com/ARMmbed/mbed-os-examples-docs_only

juhaylinen replied:
Do you know what are the plans for other USB device examples? I can add the example if needed.

AnotherButler replied:
We would like to make the other USB device examples importable examples instead of having them hardcoded in the documentation.

juhaylinen replied:
PR created https://github.com/ARMmbed/mbed-os-examples-docs_only/pull/50

AnotherButler commented on docs/api/usb/USBCDC_ECM.md:
```diff
@@ -0,0 +1,52 @@
+
+## USBCDC_ECM class reference
+
+[![View code](https://www.mbed.com/embed/?type=library)](https://os.mbed.com/docs/mbed-os/development/mbed-os-api-doxy/class_u_s_b_c_d_c___e_c_m.html)
```
I'll wait to approve this PR until the code merges, and I can check this link.

AnotherButler commented on docs/api/usb/USBCDC_ECM.md:
```diff
@@ -0,0 +1,52 @@
+# USBCDC_ECM
+
+<span class="images">![](https://os.mbed.com/docs/mbed-os/development/mbed-os-api-doxy/class_u_s_b_c_d_c___e_c_m.png)<span>USBCDC_ECM class hierarchy</span></span>
```
I'll wait to approve this PR until the code merges, and I can check this link.

AnotherButler reviewed (commented):
Thanks for the PR. Please review my changes and address my queries.

AnotherButler commented:
Does anyone know why this is failing Travis?

cmonr commented:
@AnotherButler `apt-get update` is notoriously flaky.

Restarting the job in this instance is sufficient, but I would suggest having someone refactor `.travis.yml` at some point to remove the command instance.

cmonr commented:
And naturally both @geky, @kegilbert, and I will be out until next month :grin:

Commit:
Add more context to USBCDC_ECM.md

Add use case context provided from engineering in comments.

kegilbert commented:
https://github.com/ARMmbed/mbed-os-5-docs/blob/development/.travis.yml#L22 this apt-get update? That was added for for the code snippet compilation validation. Looks like it has a safe-guard somewhat in place, but @geky would know more.

0xc0170 commented:
@c1728p9 Can you please review?

0xc0170 commented:
> Documentation for PR ARMmbed/mbed-os#9443

About to get integrated, for 5.13. Lets go ahead with this one to have it docs

AnotherButler commented:
If I merge this before the Doxygen pages exist, it'll break my build. I'll merge as soon as the Doxygen's ready.

mikter commented:
@AnotherButler and @c1728p9  are we ok to merge the documentation and example of the USB class pull requests. 
https://github.com/ARMmbed/mbed-os-examples-docs_only/pull/50

AnotherButler commented:
I believe @0xc0170 requested changes on ARMmbed/mbed-os-examples-docs_only#50.
I can't merge this documentation until the Doxygen appears. If I merge with an embedded link to a Doxygen page that doesn't exist, it breaks my build, and no Mbed OS docs will be able to build.

AnotherButler commented:
The Doxygen has merged 🎉 We only need the example changes, and then this is good to go

Commit:
Fix example format in USBCDC_ECM.md

Replace hardcoded example with transcluded rendered example, so users can import it directly into the Arm Online Compiler.

AnotherButler reviewed (approved):
Thanks so much for all of your hard work on this PR 👍 It looks great.
